### DISCUSSION TITLE
Why doesn't AFL++ support Windows? How feasible is a solid windows port?

### DISCUSSION THREAD

**Discussion by eddsanity** [AFLplusplus/AFLplusplus#1556] [General]
I am newly exposed to AFL/AFL++ and I was wondering why it lacked Windows support. I was also wondering about the feasibility of a Windows port as a contribution or something?

**Comment by domenukk** [comment]
The main roadblock is that AFL traditionally relied heavily on a forkserver, which forks the target for every execution. Windows simply doesn't support `fork`. Anything else, specifically persistent mode, could be ported.
You can instead also take a look at [LibAFL](https://github.com/AFLplusplus/LibAFL) which already supports Windows out of the box

**Reply by domenukk** [reply]
One additional note, you can already fuzz Windows applications with afl++, either by placing the target inside a VM and using `nyx` mode, or using the qemu wine mode (`-W`) - both with certain complexities and drawbacks, of course.

**Reply by eddsanity** [reply]
I may be missing something, but doesn't Windows have its own `ZwCreateProcess()` function that, as far as I am aware, forks?

**Reply by eddsanity** [reply]
Also, I am not sure I understand what drawbacks would be (though I am sure there are some), would you mind elaborating?

**Comment by vanhauser-thc** [comment]
Note that there is also https://github.com/googleprojectzero/winafl and afl port to windows.
There was a plan to port alf++ to Windows too, but I have zero time currently and there would be 2 other things of higher priority when I finally have time again.
